bd-resilient-kv/src/kv_journal/in_memory.rs

```diff
@@ -0,0 +1,446 @@
+  /// invalid.
+  pub fn from_buffer(
+    buffer: &'a mut [u8],
+    high_water_mark_ratio: Option<f32>,
+    callback: Option<HighWaterMarkCallback>,
+  ) -> anyhow::Result<Self> {
+    if buffer.len() < 16 {
+      anyhow::bail!(
+        "Buffer too small: {} bytes, need at least 16 bytes for header",
+        buffer.len()
+      );
+    }
+
+    // Validate high water mark ratio
+    let ratio = Self::validate_high_water_mark_ratio(high_water_mark_ratio)?;
+
+    let version_bytes: [u8; 8] = buffer[.. 8]
+      .try_into()
+      .map_err(|_| anyhow::anyhow!("Failed to read version bytes"))?;
+
+    let position_bytes: [u8; 8] = buffer[8 .. 16]
+      .try_into()
+      .map_err(|_| anyhow::anyhow!("Failed to read position bytes"))?;
+
+    let position_value = u64::from_le_bytes(position_bytes);
+    let position_usize = usize::try_from(position_value)
+      .map_err(|_| anyhow::anyhow!("Position value too large: {position_value}"))?;
+
+    // Calculate high water mark position
+    #[allow(
+      clippy::cast_precision_loss,
+      clippy::cast_possible_truncation,
+      clippy::cast_sign_loss
+    )]
+    let high_water_mark = (buffer.len() as f32 * ratio) as usize;
+
+    let version = u64::from_le_bytes(version_bytes);
+    if version != VERSION {
+      anyhow::bail!("Unsupported version: {}, expected {}", version, VERSION);
+    }
+    if position_usize >= buffer.len() {
+      anyhow::bail!(
+        "Invalid position: {}, buffer size: {}",
+        position_usize,
+        buffer.len()
+      );
+    }
+
+    // Extract timestamp from existing metadata
+    let init_timestamp = Self::extract_timestamp_from_buffer(&buffer[16 .. position_usize])?;
+
+    // Create final struct after all validation is complete
+    Ok(Self {
+      version,
+      position: position_usize,
+      buffer,
+      high_water_mark,
+      high_water_mark_callback: callback,
+      high_water_mark_triggered: position_usize >= high_water_mark,
+      initialized_at_unix_time_ns: init_timestamp,
+    })
+  }
+
+  /// Get a copy of the buffer for testing purposes
+  #[cfg(test)]
+  #[must_use]
+  pub fn buffer_copy(&self) -> Vec<u8> {
+    self.buffer.to_vec()
+  }
+
+  /// Validate high water mark ratio and return the validated value.
+  fn validate_high_water_mark_ratio(ratio: Option<f32>) -> anyhow::Result<f32> {
+    let ratio = ratio.unwrap_or(0.8);
+    if !(0.0 ..= 1.0).contains(&ratio) {
+      anyhow::bail!(
+        "High water mark ratio must be between 0.0 and 1.0, got: {}",
+        ratio
+      );
+    }
+    Ok(ratio)
+  }
+
+  /// Create and write metadata object with given timestamp.
+  /// This data is always written starting at offset 17 in the buffer.
+  /// Returns the new current position (17 + metadata length).
+  fn write_metadata(buffer: &mut [u8], timestamp: u64) -> anyhow::Result<usize> {
+    let buffer_len = buffer.len();
+    let mut cursor = &mut buffer[17 ..];
```

Comment:
The constructor only validates the buffer is 16 bytes I think? Can't this fail with a bad size? Should the constructor pedantically validate it's >= 17? Might be similar issues elsewhere.